nep-19.mediawiki

```diff
@@ -22,11 +24,248 @@ about parameters, variables and storage items that exists in the contract source
+Of course, the debugger supports both v1 and v2 debug information, enabling programming language teams
+to move to the new format in a time frame of their own choosing. There are no plans to remove support
+for v1 debug information, so programming language teams don't have to update their compilers at all
+if they so choose.
+
+<blockquote>
+Note, generating richer contract type model in NCCS is straightforward in part because of the rich type
+information available in C#. Supporting some aspects of the new contract type model may be more difficult
+or impossible in Neo compilers for dynamic languages such as Python. The v2 format has been designed
+with such language limitations in mind. However, the contract type model is expected to evolve based
+on feedback from Neo compiler teams as they upgrade their compilers to enable the new debugger functionality.
+</blockquote>
+
 == Rationale ==
 
-This format has been implemented by the Neo Smart Contract Debugger and multiple Neo smart contract
+Version 1 of this format has been implemented by the Neo Smart Contract Debugger and multiple Neo smart contract
 compilers including NCCS, NEON, neo-boa, Neow3j and neo-go.
 
+Version 2 of this format enables an improved variable inspection experience and supports unified type information
+for both runtime and storage items.
+
+== NeoVM Item Type Model ==
+
+Generally, type information is used to decode NeoVM types during debugging. For example, 
+NeoVM has no native string type - strings are represented in NeoVM as a
+[https://github.com/neo-project/neo-vm/blob/master/src/neo-vm/Types/ByteString.cs ByteString].
+There is no way to tell that a given ByteString represents a string or not. However, with type
+information about a runtime or storage type available from the debug info file, the debugger can
+tell if a ByteString actually represents some higher order type such as a string, and address 
+or even a serialized compound type.
+
+The primary difference between v1 and v2 of the debug info format is the way type information is encoded.
+Version 1 uses the existing [https://github.com/neo-project/neo/blob/master/src/neo/SmartContract/ContractParameterType.cs ContractParameterType enum type]
+while Version 2 defines a type hierarchy (with string encoding) to represent additional type
+information that would be impossible to encode with ContractParameterType.
+
+=== ContractType Type Model (Version 2) ===
+
+This section describes both the abstract ContractType model as well as the concrete syntax used to 
+string encoding ContractTypes.
+
+Note, because the ContractType model namespace is unbounded (i.e. contract developers can define
+their own named structs), built-in ContractTypes use a '''#''' prefix when serialized. This eliminates
+the possibility of a name collision since '''#''' is not a supported identifier character in most
+programming languages.
+
+The Version 2 type model includes six concrete ContractTypes. 
```

Comment:
Need to add a mechanism to encode Iterators - probably as an intrinsic type (i.e. #Iterator).